lib/service/connect.go

```diff
@@ -684,6 +687,114 @@ func (process *TeleportProcess) firstTimeConnect(role types.SystemRole) (*Connec
 	return connector, nil
 }
 
+func (process *TeleportProcess) initOpenSSH() {
+	process.RegisterWithAuthServer(types.RoleNode, SSHIdentityEvent)
+	process.SSHd = openssh.NewSSHD()
+	process.RegisterCriticalFunc("common.rotate_openssh", process.syncOpenSSHRotationState)
+}
+
+func (process *TeleportProcess) syncOpenSSHRotationState() error {
+	if _, err := process.WaitForEvent(process.GracefulExitContext(), TeleportReadyEvent); err != nil {
+		return nil
+	}
+	conn, err := process.WaitForConnector(SSHIdentityEvent, nil)
+	if conn == nil {
+		return trace.Wrap(err)
+	}
+	defer conn.Close()
+
+	ctx := context.Background()
+
+	status, err := process.syncRotationStateAndBroadcast(conn)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	id, err := process.storage.ReadIdentity(auth.IdentityCurrent, types.RoleInstance)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	cas, err := conn.Client.GetCertAuthorities(ctx, types.OpenSSHCA, false)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	keysDir := process.Config.OpenSSH.KeysDir
+	if err := openssh.TryCreateDefaultAgentlesKeysDir(keysDir); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := openssh.WriteKeys(keysDir, id, cas); err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = process.SSHd.UpdateConfig(openssh.SSHdConfigUpdate{
+		SSHdConfigPath:         process.Config.OpenSSH.SSHDConfigPath,
+		TeleportSSHDConfigPath: openssh.AgentlessSSHDConfigPath,
+		OpenSSHCAPath:          filepath.Join(keysDir, openssh.TeleportOpenSSHCA),
+		HostKeyPaths:           []string{filepath.Join(keysDir, openssh.TeleportKey)},
+		HostCertPaths:          []string{filepath.Join(keysDir, openssh.TeleportCert)},
+	}, process.Config.OpenSSH.RestartSSHD)
+
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	state, err := process.storage.GetState(types.RoleNode)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	mostRecentRotation := state.Spec.Rotation
+	if !status.needsReload {
+		for _, ca := range cas {
+			if ca.GetRotation().LastRotated.After(mostRecentRotation.LastRotated) {
+				mostRecentRotation = ca.GetRotation()
+			}
+		}
+	}
+
+	if err := process.storage.WriteState(types.RoleNode, *state); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := registerServer(process.Config, ctx, conn.Client, mostRecentRotation); err != nil {
+		return trace.Wrap(err)
+	}
+
+	go func() {
+		process.Shutdown(ctx)
```

Comment:
Took me a sec to figure out that this is all within `RegisterCriticalFunc` so if we exit with a non-nil error the process dies anyway; please add a comment about that.